src/PathOrderMonotonic.h

```diff
@@ -0,0 +1,449 @@
+            const coord_t a_projection_min = std::min(a_start_projection, a_end_projection); //The projection of a path is the endpoint furthest back of the two endpoints.
+            const coord_t a_projection_max = std::max(a_start_projection, a_end_projection); //But in case of ties, the other endpoint counts too. Important for polylines where multiple endpoints have the same position!
+
+            const coord_t b_start_projection = dot(b->converted->front(), monotonic_vector);
+            const coord_t b_end_projection = dot(b->converted->back(), monotonic_vector);
+            const coord_t b_projection_min = std::min(b_start_projection, b_end_projection);
+            const coord_t b_projection_max = std::max(b_start_projection, b_end_projection);
+
+            return a_projection_min < b_projection_min || (a_projection_min == b_projection_min && a_projection_max < b_projection_max);
+        });
+
+        //Now that we have the segments of overlapping lines, and know in which order to print the segments, print segments in monotonic order.
+        Point current_pos = this->start_point;
+        for(Path* line : starting_lines_monotonic)
+        {
+            optimizeClosestStartPoint(*line, current_pos);
+            reordered.push_back(*line); //Plan the start of the sequence to be printed next!
+            auto connection = connections.find(line);
+            while(connection != connections.end() && starting_lines.find(connection->second) == starting_lines.end()) //Stop if the sequence ends or if we hit another starting point.
+            {
+                line = connection->second;
+                optimizeClosestStartPoint(*line, current_pos);
+                reordered.push_back(*line); //Plan this line in, to be printed next!
+                connection = connections.find(line);
+            }
+        }
+
+        std::swap(reordered, this->paths); //Store the resulting list in the main paths.
+    }
+
+protected:
+    /*!
+     * The direction in which to print montonically, encoded as vector of length
+     * 1000.
+     *
+     * The resulting ordering will cause clusters of paths to be sorted
+     * according to their projection on this vector.
+     */
+    Point monotonic_vector;
+
+    /*!
+     * Maximum distance at which lines are considered to be adjacent.
+     *
+     * The monotonicity constraint is only held for lines that are closer than
+     * this distance together.
+     */
+    coord_t max_adjacent_distance;
+
+    /*!
+     * For a given path, make sure that it is configured correctly to start
+     * printing from the best endpoint.
+     *
+     * This changes the path's ``start_vertex`` and ``backwards`` fields, and
+     * also adjusts the \ref current_pos in-place.
+     *
+     * If the path already had a ``start_vertex`` set, this will not be
+     * adjusted. Only the ``current_pos`` will be set then.
+     *
+     * Will cause a crash if given a path with 0 vertices!
```

Comment:
We do that check here: https://github.com/Ultimaker/CuraEngine/pull/1476/files#diff-4441e292f911dd779107cb41ee83900436da2df15d3830ab0c9dcdfcdbd48b2eR101

Since this is an internal function, we don't need to re-check all edge cases within every nested function call. This documentation is there to hopefully ensure that whoever maintains this later will make sure that empty paths are filtered out before it gets to this internal function.

Reply:
Fair enough.